Flaky Test – io.camunda.zeebe.engine.processing.processinstance.CreateProcessInstanceAnywhereTest#shouldSubscribeToScopeEvents

**Summary**

> Try to answer the following as best as possible

- How often does the test fail? once thus far
- Does it block your work? no, but we should fix it while we are working on this topic
- Do we suspect that it is a real failure? Possibly

**Failures**

> Outline known failure cases, e.g. a failed assertion and its stacktrace obtained from Jenkins

<details><summary>Example assertion failure</summary>
<pre>
java.lang.AssertionError: 
Expecting actual to contain the specified subsequence but failed to find the element at subsequence index 1 in actual:
subsequence element not found in actual:
  ("timer-boundary-event", BOUNDARY_EVENT, ELEMENT_COMPLETED)
actual:
  [("process", PROCESS, ELEMENT_ACTIVATING),
    ("process", PROCESS, ELEMENT_ACTIVATED),
    ("subprocess", SUB_PROCESS, ELEMENT_ACTIVATING),
    ("subprocess", SUB_PROCESS, ELEMENT_ACTIVATED),
    ("task", SERVICE_TASK, ACTIVATE_ELEMENT),
    ("task", SERVICE_TASK, ELEMENT_ACTIVATING),
    ("task", SERVICE_TASK, ELEMENT_ACTIVATED),
    ("message-boundary-event", BOUNDARY_EVENT, ELEMENT_ACTIVATING),
    ("message-boundary-event", BOUNDARY_EVENT, ELEMENT_ACTIVATED),
    ("message-boundary-event", BOUNDARY_EVENT, COMPLETE_ELEMENT),
    ("message-boundary-event", BOUNDARY_EVENT, ELEMENT_COMPLETING),
    ("message-boundary-event", BOUNDARY_EVENT, ELEMENT_COMPLETED),
    ("sequenceFlow_e69dac8f-c702-4d2e-bbcd-14f2f1847305", SEQUENCE_FLOW, SEQUENCE_FLOW_TAKEN),
    ("endEvent_40366a6a-d5ed-41a4-b5ca-69c48b1fab96", END_EVENT, ACTIVATE_ELEMENT),
    ("endEvent_40366a6a-d5ed-41a4-b5ca-69c48b1fab96", END_EVENT, ELEMENT_ACTIVATING),
    ("endEvent_40366a6a-d5ed-41a4-b5ca-69c48b1fab96", END_EVENT, ELEMENT_ACTIVATED),
    ("endEvent_40366a6a-d5ed-41a4-b5ca-69c48b1fab96", END_EVENT, ELEMENT_COMPLETING),
    ("endEvent_40366a6a-d5ed-41a4-b5ca-69c48b1fab96", END_EVENT, ELEMENT_COMPLETED),
    ("message-boundary-event", BOUNDARY_EVENT, ELEMENT_ACTIVATING),
    ("message-boundary-event", BOUNDARY_EVENT, ELEMENT_ACTIVATED),
    ("message-boundary-event", BOUNDARY_EVENT, COMPLETE_ELEMENT),
    ("message-boundary-event", BOUNDARY_EVENT, ELEMENT_COMPLETING),
    ("message-boundary-event", BOUNDARY_EVENT, ELEMENT_COMPLETED),
    ("sequenceFlow_e69dac8f-c702-4d2e-bbcd-14f2f1847305", SEQUENCE_FLOW, SEQUENCE_FLOW_TAKEN),
    ("endEvent_40366a6a-d5ed-41a4-b5ca-69c48b1fab96", END_EVENT, ACTIVATE_ELEMENT),
    ("task", SERVICE_TASK, COMPLETE_ELEMENT),
    ("endEvent_40366a6a-d5ed-41a4-b5ca-69c48b1fab96", END_EVENT, ELEMENT_ACTIVATING),
    ("endEvent_40366a6a-d5ed-41a4-b5ca-69c48b1fab96", END_EVENT, ELEMENT_ACTIVATED),
    ("endEvent_40366a6a-d5ed-41a4-b5ca-69c48b1fab96", END_EVENT, ELEMENT_COMPLETING),
    ("endEvent_40366a6a-d5ed-41a4-b5ca-69c48b1fab96", END_EVENT, ELEMENT_COMPLETED),
    ("task", SERVICE_TASK, ELEMENT_COMPLETING),
    ("task", SERVICE_TASK, ELEMENT_COMPLETED),
    ("sequenceFlow_3240c7af-6690-4a8c-a081-a5bcbca48861", SEQUENCE_FLOW, SEQUENCE_FLOW_TAKEN),
    ("endEvent_4b4a8c5a-3bbb-454e-8b10-73c13a5d8824", END_EVENT, ACTIVATE_ELEMENT),
    ("endEvent_4b4a8c5a-3bbb-454e-8b10-73c13a5d8824", END_EVENT, ELEMENT_ACTIVATING),
    ("endEvent_4b4a8c5a-3bbb-454e-8b10-73c13a5d8824", END_EVENT, ELEMENT_ACTIVATED),
    ("endEvent_4b4a8c5a-3bbb-454e-8b10-73c13a5d8824", END_EVENT, ELEMENT_COMPLETING),
    ("endEvent_4b4a8c5a-3bbb-454e-8b10-73c13a5d8824", END_EVENT, ELEMENT_COMPLETED),
    ("subprocess", SUB_PROCESS, COMPLETE_ELEMENT),
    ("subprocess", SUB_PROCESS, ELEMENT_COMPLETING),
    ("subprocess", SUB_PROCESS, ELEMENT_COMPLETED),
    ("sequenceFlow_dadb18b0-0ecb-41b8-a684-812d3ec049e7", SEQUENCE_FLOW, SEQUENCE_FLOW_TAKEN),
    ("endEvent_62421b46-e382-465c-9ef7-bbef8fc78aae", END_EVENT, ACTIVATE_ELEMENT),
    ("endEvent_62421b46-e382-465c-9ef7-bbef8fc78aae", END_EVENT, ELEMENT_ACTIVATING),
    ("endEvent_62421b46-e382-465c-9ef7-bbef8fc78aae", END_EVENT, ELEMENT_ACTIVATED),
    ("endEvent_62421b46-e382-465c-9ef7-bbef8fc78aae", END_EVENT, ELEMENT_COMPLETING),
    ("endEvent_62421b46-e382-465c-9ef7-bbef8fc78aae", END_EVENT, ELEMENT_COMPLETED),
    ("process", PROCESS, COMPLETE_ELEMENT),
    ("process", PROCESS, ELEMENT_COMPLETING),
    ("process", PROCESS, ELEMENT_COMPLETED)]
subsequence:
  [("message-boundary-event", BOUNDARY_EVENT, ELEMENT_COMPLETED),
    ("timer-boundary-event", BOUNDARY_EVENT, ELEMENT_COMPLETED),
    ("process", PROCESS, ELEMENT_COMPLETED)]
	at io.camunda.zeebe.engine.processing.processinstance.CreateProcessInstanceAnywhereTest.shouldSubscribeToScopeEvents(CreateProcessInstanceAnywhereTest.java:778)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.apache.maven.surefire.junitplatform.LazyLauncher.execute(LazyLauncher.java:55)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.lambda$execute$1(JUnitPlatformProvider.java:234)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:228)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:175)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:131)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:456)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:169)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:595)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:581)
</pre>
</details>

**Hypotheses**

It seems like the message that is send gets correlated twice, messing up the assertion that waits for the timer boundary event to be triggered. Maybe there is multiple message subscriptions created?

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/CreateProcessInstanceAnywhereTest.java
@@ -13,6 +13,7 @@ import static org.assertj.core.groups.Tuple.tuple;
 import io.camunda.zeebe.engine.util.EngineRule;
 import io.camunda.zeebe.model.bpmn.Bpmn;
 import io.camunda.zeebe.protocol.record.Record;
+import io.camunda.zeebe.protocol.record.intent.MessageSubscriptionIntent;
 import io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent;
 import io.camunda.zeebe.protocol.record.intent.ProcessMessageSubscriptionIntent;
 import io.camunda.zeebe.protocol.record.intent.TimerIntent;
@@ -641,11 +642,15 @@ public class CreateProcessInstanceAnywhereTest {
             .create();
 
     // when
-    RecordingExporter.processMessageSubscriptionRecords(ProcessMessageSubscriptionIntent.CREATED)
+    RecordingExporter.messageSubscriptionRecords(MessageSubscriptionIntent.CREATED)
         .withProcessInstanceKey(processInstanceKey)
         .await();
 
     ENGINE.message().withName(messageName).withCorrelationKey("key-1").publish();
+
+    RecordingExporter.messageSubscriptionRecords(MessageSubscriptionIntent.CORRELATED)
+        .withProcessInstanceKey(processInstanceKey)
+        .await();
 
     RecordingExporter.timerRecords(TimerIntent.CREATED)
         .withProcessInstanceKey(processInstanceKey)
@@ -733,11 +738,15 @@ public class CreateProcessInstanceAnywhereTest {
             .create();
 
     // when
-    RecordingExporter.processMessageSubscriptionRecords(ProcessMessageSubscriptionIntent.CREATED)
+    RecordingExporter.messageSubscriptionRecords(MessageSubscriptionIntent.CREATED)
         .withProcessInstanceKey(processInstanceKey)
         .await();
 
     ENGINE.message().withName(messageName).withCorrelationKey("key-1").publish();
+
+    RecordingExporter.messageSubscriptionRecords(MessageSubscriptionIntent.CORRELATED)
+        .withProcessInstanceKey(processInstanceKey)
+        .await();
 
     RecordingExporter.timerRecords(TimerIntent.CREATED)
         .withProcessInstanceKey(processInstanceKey)

